tools/kubernetes/migration.py

```diff
@@ -0,0 +1,187 @@
+        
+        response_data = response.json()
+        pipelines.extend(response_data.get("pipelines", []))     
+        page_token = response_data.get("next_page_token")
+        if not page_token:
+            break
+
+    return pipelines
+
+# Fetch all versions for a given pipeline
+def fetch_pipeline_versions(kfp_server_host, pipeline_id, headers, verify, namespace, batch_size):
+    versions = []
+    page_token = ""
+
+    while True:
+        url = (
+            f"{kfp_server_host}/apis/v2beta1/pipelines/{pipeline_id}/versions?"
+            f"sort_by=created_at&order_by=asc&namespace={namespace}&page_size={batch_size}"
+        )
+        if page_token:
+            url += f"&page_token={page_token}"
+
+        response = requests.get(url, headers=headers, verify=verify, timeout=DEFAULT_TIMEOUT)
+        if response.status_code != 200:
+            raise Exception(f"Error fetching versions for pipeline {pipeline_id}: {response.status_code} - {response.text}")
+        
+        response_data = response.json()
+        # print(f"Response data for version:{response_data}")
+        versions.extend(response_data.get("pipeline_versions", [])) 
+        page_token = response_data.get("next_page_token")
+        if not page_token:
+            break
+
+    return versions
+
+# Creates a Kubernetes-safe name for a pipeline version by combining the pipeline name and version ID.
+# If the name is too long, it shortens it and adds index at the end to keep it unique.
+def get_version_name(pipeline_name, version_display_name, index, add_prefix):
+    if add_prefix:
+       original_version_name = f"{pipeline_name}-{version_display_name}"
+    else:
+        original_version_name = f"{version_display_name}"    
+    
+    # Clean and convert the pipeline name to be k8s-compatible
+    formatted_name = utils.maybe_rename_for_k8s(original_version_name)
+
+    if len(formatted_name) > MAX_NAME_LENGTH:
+        suffix = f"-{index}"
+        max_name_len = MAX_NAME_LENGTH - len(suffix)
+        trimmed_name = original_version_name[:max_name_len]
+        formatted_name = utils.maybe_rename_for_k8s(trimmed_name + suffix)
+
+    return formatted_name
+
+# Convert a pipeline and its versions into k8s format
+def convert_to_k8s_format(pipeline, pipeline_versions, add_prefix, namespace):
+    k8s_objects = []
+
+    # Use namespace from pipeline object if available; otherwise fallback to default
+    namespace = pipeline.get('namespace') or namespace
+
+    # Clean and convert the pipeline name to be k8s-compatible
+    pipeline_name = utils.maybe_rename_for_k8s(pipeline.get('display_name', 'unnamed'))
+    versions = pipeline_versions
+
+    # Create the Pipeline object
+    pipeline_obj = {
+        "apiVersion": K8S_PIPELINE_API_VERSION,
+        "kind": "Pipeline",
+        "metadata": {
+            "name": pipeline_name,
+            "namespace": namespace,
```

Comment:
Sorry for not thinking about this earlier, but it'd be nice if you put the original display name and original UIDs as annotations on each pipeline and pipeline version.

For example:

```yaml
annotations:
  pipelines.kubeflow.org/original-id: "UID-HERE"
  pipelines.kubeflow.org/display-name: "Display name here"
```